apps/prairielearn/src/models/enrollment.ts

```diff
@@ -443,63 +521,82 @@ export async function selectAndLockEnrollmentById(id: string) {
  * The function will lock the enrollment row and create an audit event based on the status change.
  */
 export async function setEnrollmentStatus({
-  enrollment_id,
-  agent_user_id,
-  agent_authn_user_id,
+  enrollment,
   status,
-  required_status,
+  authzData,
+  requestedRole,
 }: {
-  enrollment_id: string;
-  agent_user_id: string | null;
-  agent_authn_user_id: string | null;
-  status: 'rejected' | 'blocked' | 'removed';
-  required_status: 'invited' | 'joined';
+  enrollment: Enrollment;
+  status: 'rejected' | 'blocked' | 'removed' | 'joined';
+  authzData: AuthzData;
+  requestedRole: CourseInstanceRole;
 }): Promise<Enrollment> {
-  return await runInTransactionAsync(async () => {
-    const oldEnrollment = await selectAndLockEnrollmentById(enrollment_id);
-    if (oldEnrollment.user_id) {
-      await selectAndLockUserById(oldEnrollment.user_id);
+  const transitionInformation: {
+    previousStatus: EnumEnrollmentStatus;
+    actionDetail: SupportedActionsForTable<'enrollments'>;
+    allowedRoles: CourseInstanceRole[];
+  } = run(() => {
+    switch (status) {
+      case 'joined':
+        return {
+          previousStatus: 'blocked',
+          actionDetail: 'unblocked',
+          allowedRoles: ['Student Data Viewer', 'Student Data Editor'],
+        };
+      case 'removed':
+        return {
+          previousStatus: 'joined',
+          actionDetail: 'removed',
+          allowedRoles: ['Student'],
+        };
+      case 'rejected':
+        return {
+          previousStatus: 'invited',
+          actionDetail: 'invitation_rejected',
+          allowedRoles: ['Student'],
+        };
+      case 'blocked':
+        return {
+          previousStatus: 'joined',
+          actionDetail: 'blocked',
+          allowedRoles: ['Student Data Viewer', 'Student Data Editor'],
+        };
+      default:
+        assertNever(status);
     }
+  });
 
+  return await runInTransactionAsync(async () => {
+    const lockedEnrollment = await _selectAndLockEnrollment(enrollment.id);
+    if (lockedEnrollment.user_id) {
+      await selectAndLockUser(lockedEnrollment.user_id);
+    }
     // The enrollment is already in the desired status, so we can return early.
-    if (oldEnrollment.status === status) {
-      return oldEnrollment;
+    if (lockedEnrollment.status === status) {
+      return lockedEnrollment;
     }
 
-    if (oldEnrollment.status !== required_status) {
-      throw new Error(
-        `Enrollment is not in the required status. Expected ${required_status}, but got ${oldEnrollment.status}`,
-      );
-    }
+    // Assert that the enrollment is in the previous status.
+    assertEnrollmentStatus(enrollment, transitionInformation.previousStatus);
+
+    // Assert that the caller is authorized to perform the action.
+    assertHasRole(authzData, requestedRole, transitionInformation.allowedRoles);
```

Comment:
This is the case that I was trying (and probably failing) to articulate: the reason that I thought `assertHasRole` would still need an optional third argument.